File: src/main/java/me/lokka30/levelledmobs/listeners/EntityPickupItemListener.java
package me.lokka30.levelledmobs.listeners;

import me.lokka30.levelledmobs.LevelledMobs;
import me.lokka30.levelledmobs.misc.PickedUpEquipment;
import me.lokka30.levelledmobs.wrappers.LivingEntityWrapper;
import me.lokka30.levelledmobs.wrappers.SchedulerWrapper;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityPickupItemEvent;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

public class EntityPickupItemListener implements Listener {
    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onEntityPickupItemEvent(final @NotNull EntityPickupItemEvent event){
        // sorry guys this is a Paper only feature
        if (!LevelledMobs.getInstance().getVerInfo().getIsRunningPaper()) return;
        if (event.getEntity() instanceof Player) return;

        final LivingEntityWrapper lmEntity = LivingEntityWrapper.getInstance(
                event.getEntity(), LevelledMobs.getInstance());

        if (!lmEntity.isLevelled() || lmEntity.getLivingEntity().getEquipment() == null){
            lmEntity.free();
            return;
        }

        // if you don't clone the item then it will change to air in the next function
        final ItemStack itemStack = event.getItem().getItemStack().clone();
        final PickedUpEquipment pickedUpEquipment = new PickedUpEquipment(lmEntity);
        final SchedulerWrapper wrapper = new SchedulerWrapper(lmEntity.getLivingEntity(),
                () -> pickedUpEquipment.checkEquipment(itemStack));
        wrapper.runDelayed(1L);
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/listeners/EntityPickupItemListener.kt
package io.github.arcaneplugins.levelledmobs.listeners

import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.misc.PickedUpEquipment
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper
import io.github.arcaneplugins.levelledmobs.wrappers.SchedulerWrapper
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityPickupItemEvent

/**
 * Listens for when an entity picks up items and tracks the items only if
 * the entity equipped the items so that later we can be sure not to
 * destroy these items
 *
 * @author stumper66
 * @since 3.14.0
 */
class EntityPickupItemListener : Listener {
    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onEntityPickupItemEvent(event: EntityPickupItemEvent) {
        // sorry guys this is a Paper only feature
        if (!LevelledMobs.instance.ver.isRunningPaper) return
        if (event.entity is Player) return

        val lmEntity = LivingEntityWrapper.getInstance(event.entity)

        if (!lmEntity.isLevelled || lmEntity.livingEntity.equipment == null) {
            lmEntity.free()
            return
        }

        // if you don't clone the item then it will change to air in the next function
        val itemStack = event.item.itemStack.clone()
        val pickedUpEquipment = PickedUpEquipment(lmEntity)
        val wrapper = SchedulerWrapper(lmEntity.livingEntity) {
            pickedUpEquipment.checkEquipment(itemStack)
            lmEntity.free()
        }

        lmEntity.inUseCount.getAndIncrement()
        wrapper.runDelayed(1L)
    }
}